compiler/src/iree/compiler/Dialect/Stream/Transforms/AnnotateConstantTransientSize.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler::IREE::Stream {
+
+#define GEN_PASS_DEF_ANNOTATECONSTANTTRANSIENTSIZEPASS
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h.inc"
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// --iree-stream-annotate-constant-transient-size
+//===----------------------------------------------------------------------===//
+
+// Merges a new named attribute into an existing reflection dictionary.
+static DictionaryAttr
+getReflectionAttrWithNamedAttr(DictionaryAttr existingAttr,
+                               NamedAttribute namedAttr) {
+  SmallVector<NamedAttribute> reflectionAttrs;
+  if (existingAttr) {
+    llvm::append_range(reflectionAttrs, existingAttr.getValue());
+  }
+  reflectionAttrs.push_back(namedAttr);
+  return DictionaryAttr::get(namedAttr.getName().getContext(), reflectionAttrs);
+}
+
+// Analyzes a size query function to detect if it returns a constant value.
+// Returns the constant IntegerAttr if detected, otherwise std::nullopt.
+// Uses pattern matching to support all sources of constant values.
+static std::optional<IntegerAttr>
+detectConstantReturn(IREE::Util::FuncOp funcOp) {
+  // Must have exactly one block (no control flow).
+  if (!funcOp.getBody().hasOneBlock()) {
+    return std::nullopt;
+  }
+
+  // Find the return operation.
+  Block &block = funcOp.getBody().front();
+  auto returnOp = dyn_cast<IREE::Util::ReturnOp>(block.getTerminator());
+  if (!returnOp || returnOp.getNumOperands() != 1) {
+    return std::nullopt;
+  }
+
+  // Use pattern matching to detect constant return value.
+  IntegerAttr constantValue;
+  if (matchPattern(returnOp.getOperand(0), m_Constant(&constantValue))) {
+    return constantValue;
+  }
+
+  return std::nullopt;
+}
+
+struct AnnotateConstantTransientSizePass
```

Comment:
I dont see this pass being run anywhere? Is this supposed to be run by the caller somehow?

Reply:
good catch! cherry pick bug 🤦